Avoid linking in libnuma.so if running on a non-NUMA system

lpereira opened the pull request:
In an effort to reduce the system calls performed during the initialization, this seemed like another low-hanging fruit: only link in libnuma.so if the system has more than one NUMA node.

A NUMA-capable system will have something like "0-3\n" in "/sys/devices/system/node/possible" if there are 4 nodes that could potentially come up.  A non-NUMA system will have only "0\n".

lpereira committed:
Avoid linking in libnuma.so if running on a non-NUMA system

In an effort to reduce the system calls performed during the initialization,
this seemed like another low-hanging fruit: only link in libnuma.so if the
system has more than one NUMA node.

jkotas commented:
How many syscalls does this save?

janvorli commented:
I don't think we should add this complexity to our code. If someone doesn't have NUMA capable device, it is enough to not to install the libnuma package and nothing will be loaded into the process. That seems like enough control for people that would care.
Moreover the libnuma.so is only 43kB large.

lpereira commented:
> How many syscalls does this save?

Saves [~100 syscalls](https://gist.github.com/lpereira/15fe557107435f1c57cb7a57621e436a) (50 in libcoreclr.so, and 50 in libcorejit.so).

Look at the timestamps, too: it's about 0.1s per time libnuma.so is linked in, so around 0.2s shaved off from initialization.  (Mind you that these timestamps are exaggerated given this was running under strace capturing the stack trace in addition to the syscalls; but this isn't insignificant anyway.)

lpereira commented:
> I don't think we should add this complexity to our code. If someone doesn't have NUMA capable device, it is enough to not to install the libnuma package and nothing will be loaded into the process. That seems like enough control for people that would care.

Many distributions install the library by default, especially the "server" variants -- even if running on non-NUMA systems.

janvorli commented on src/gc/unix/gcenv.unix.cpp:
```diff
@@ -157,6 +160,60 @@ FOR_ALL_NUMA_FUNCTIONS
 #undef PER_FUNCTION_BLOCK
 #endif // HAVE_NUMA_H
 
+#if defined(__linux__) && HAVE_NUMA_H
```
A nit - if you moved the `#endif` above after the end of your added code, this `#if defined(__linux__) && HAVE_NUMA_H` / `#elif HAVE_NUMA_H` would be a bit simpler - just `#ifdef __linux__` / `#else`.

lpereira replied:
Done.

janvorli commented on src/gc/unix/gcenv.unix.cpp:
```diff
@@ -157,6 +160,60 @@ FOR_ALL_NUMA_FUNCTIONS
+        if (!strncmp(buffer, "0\n", bytesRead))
+        {
+            // Only one node, so avoid linking libnuma.so.
+            close(fd);
```
A nit - could you please have just a single close(fd)? E.g. by moving it before the `if`.

lpereira replied:
Done. Also fixed the case where bytesRead could be -1 with any other errno (which would be passed to strncmp()).

janvorli commented:
> Look at the timestamps, too: it's about 0.1s per time libnuma.so is linked in, so around 0.2s shaved off from initialization.

Wow, that's a lot. Ok, it makes a lot of sense then.

janvorli commented on src/pal/src/numa/numa.cpp:
```diff
@@ -46,11 +46,69 @@ bool g_numaAvailable = false;
+
+        if (fd == -1)
+        {
+            if (errno == EINTR)
```
I wonder - can this open really fail with EINTR? We have several places in PAL where we use open and we don't check for EINTR anywhere. The Linux doc says that you can get EINTR when the open would be blocked waiting to complete an open of a slow device.

lpereira replied:
Yeah, it can happen. In this case, we can just ignore this error and try use libnuma anyway.

janvorli reviewed (approved):
LGTM modulo the nits.

jkotas commented:
> Mind you that these timestamps are exaggerated given this was running under strace

Could you please measure actual non-exaggerated improvement, without running strace?

mikedn commented:
> and 50 in libcorejit.so

Hmm, why is the JIT involved in this?

janvorli commented:
The jit has its own copy of PAL.

mikedn commented:
That seems unfortunate, JIT's dependencies on OS/CRT should be pretty low. Perhaps it shouldn't use the PAL at all.

janvorli commented:
> That seems unfortunate

I agree. There were already discussions in the past that it would be good for JIT to not depend on PAL and use standard C/C++ stuff instead. I believe this issue was supposed to track that:
https://github.com/dotnet/coreclr/issues/4126

mikedn commented:
Ah, right, forgot about that one. Unfortunately the JIT has a bigger dependency on PAL than I thought - `RaiseException`, `PAL_TRY` & friends. That seems difficult to remove.

jkotas commented:
`RaiseException`, `PAL_TRY` & friends should not be that hard to remove and replace with C++ exceptions.

The whole thing should be "just work" to do (ideally, one piece at a time). I do not think there is anything that is a rocket science to solve.

mikedn commented:
If using C++ exceptions is an option then yes, it shouldn't be difficult. For some reason I had the impression that C++ exceptions are avoided in the runtime.

lpereira committed:
No need to resolve `numa_allocate_cpumask()` when linking libnuma

The result of dlsym() is never attributed to any variable, and this
function is never called, so it's safe to remove these dlsym() calls.

lpereira commented:
Measurements, as requested (using `clock_gettime(CLOCK_THREAD_CPUTIME_ID)`). Before:

**Measurement**|**Cold Cache**|**Hot Cache**
:-----:|:-----:|:-----:
1|458999ns|770614ns
2|382807ns|694449ns
3|329394ns|715913ns
4|339150ns|686243ns

And after:

**Measurement**|**Cold Cache**|**Hot Cache**
:-----:|:-----:|:-----:
1|139001ns|80824ns
2|79711ns|42391ns
3|72451ns|44044ns
4|59344ns|92955ns  

So, on this computer with little-to-no-load, and fast SSDs, it goes from around 0.5ms (avg) to around .07ms (avg). Not a lot.

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

maryamariyan commented:
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.
